### PR TITLE
fix(repo/config-migration): error fetching config file when repo cache is enabled

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Use `platform.readLocalFile` to fetch repo config file.
currently we are using `fs.getLocalFile` which fails when `repoCache=enabled`. which in turn causes migration PR to not be created/autoclosed.

```
DEBUG: MigratedDataFactory.getAsync() Error initializing renovate MigratedData
{
  "err": {
    "message": "Expected a string",
    "stack": "TypeError: Expected a string\n    at module.exports (/home/ubuntu/renovateapp/node_modules/detect-indent/index.js:134:9)\n    at Function.build (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/config-migration/branch/migrated-data.js:80:56)\n    at async Function.getAsync (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/config-migration/branch/migrated-data.js:54:26)\n    at async finaliseRepo (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/finalise/index.js:18:36)\n    at async Object.renovateRepository (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/index.js:58:13)\n    at async renovateRepository (/home/ubuntu/renovateapp/app/worker/index.js:309:26)\n    at async /home/ubuntu/renovateapp/app/worker/index.js:569:5"
  }
}
DEBUG: checkConfigMigrationBranch()
DEBUG: checkConfigMigrationBranch() Config does not need migration
...
INFO: Autoclosing PR
{
  "branchName": "renovate/migrate-config",
  "prNo": 13,
  "prTitle": "Configure Renovate"
}
DEBUG: updatePr(13, Configure Renovate - autoclosed, body)
DEBUG: PR updated
{
  "pr": 13
}
DEBUG: Deleted remote branch
{
  "branchName": "renovate/migrate-config"
}
```

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
These functions sound confusingly similar. Should we be looking deeper at why we have both and why they don't behave the same?

**Comment by Gabriel-Ladzaretti** [comment]
> These functions sound confusingly similar. Should we be looking deeper at why we have both and why they don't behave the same?

top of that, im not sure why `fs.readLocalFile` fails when repo cache is enabled though.
i mean we do clone the repository so the file should exist 🤔

**Comment by rarkins** [comment]
We shouldn't trigger a clone just for the purpose the checking migration though. This would be really wasteful.

**Comment by Gabriel-Ladzaretti** [comment]
Its not a full clone, when it comes to github for example its only one additional api call.
we are getting a json response with the raw content encoded in base64, we decode it and thats it.

**Comment by rarkins** [comment]
We don't want to be cloning or using an extra API call every run if it's only for the purpose of this migration PR and such a PR already exists and doesn't need updating. Ie I prefer to make our cache smarter instead of consuming extra GitHub calls

**Comment by rarkins** [comment]
As long as it works in both cache and non cache scenarios

**Comment by Gabriel-Ladzaretti** [comment]
> instead of saving the config to the cache, we should simply check if current branch sha is the same as before and skip the whole migration check, as the branch is already up to date.

Isnt the branch sha is changed on new commits to it? if so, we shouldnt expect too many sha changes to the migration branch.
what about comparing current parent sha to the saved parent sha in the branch cache?

**Comment by rarkins** [comment]
It's only the default branch sha which should matter, because it holds the config. And if it has changed then we would have cloned in most cases. 

Could you summarise again what exactly causes this problem? Eg is it that we haven't fully synced got on some runs, or something else?

**Comment by Gabriel-Ladzaretti** [comment]
the root of it is that this feature was written with the onboarding PR as it's model, therefore repository cache was not taken into consideration and honestly i wasnt fully aware of it yet at that time.

The current main logic is as follows:
We use the already existing `migrateConfig` to determine if the config needs migration, if it does not - null is returned (to the caller of `MigratedDataFactory.getAsync`).

In `MigratedDataFactory` we have -
https://github.com/renovatebot/renovate/blob/4d55ada172ba9c1b389d6e61a5db7d6c4287a7fe/lib/workers/repository/config-migration/branch/migrated-data.ts#L92-L96

this null indicates that the config does not need migration
https://github.com/renovatebot/renovate/blob/4d55ada172ba9c1b389d6e61a5db7d6c4287a7fe/lib/workers/repository/config-migration/branch/index.ts#L11-L19

these two function calls are invoked here -
https://github.com/renovatebot/renovate/blob/4d55ada172ba9c1b389d6e61a5db7d6c4287a7fe/lib/workers/repository/finalise/index.ts#L19-L24

so the main problem is that it is assumed that `renovate.json` file always exists, which it doesnt when repo cache is enable, so in turn `fs.readLoaclFile` fails and causes exception in `detectIndent` which causes null to be return -> which mistakably treated as if the repo does not need migration and basically the branch is getting pruned as "it shouldn't exist" and the PR is auto closed. 



so the suggested solution, as per this discussion would be to cache this branch, and check the parent sha to determine whether to try migration and fetch the raw config or not.

**Comment by Gabriel-Ladzaretti** [comment]
In the current state, we cant use repository cache in the migration flow.
This is because the cache is set before the `finalize` flow which contains the migration, therefore we cant cache the migration branch.

As i see it, we have three options:

1. Move the migration flow to the worker's main flow, just before the cache is set
2. Save a snapshot of the cache before it is being overwritten + alter the cache after it is being set in order to save the migration branch data
3. Set the repository cache in the finalize stage 

1,2 are doable, 2 requires cache manipulation. 3 i assume would break quite a few things. 

To sum it up, can we move the migration flow into the repo worker's main flow? Note that we call `renovateRepository` recursively after the cache is set.

WDYT? 

once cache is available, we can compare parent Sha in order to determine if new migration is required.  

https://github.com/renovatebot/renovate/blob/4f82df4c3d501641ae05ce713bb1328d339ec5d9/lib/workers/repository/index.ts#L40-L60

**Comment by rarkins** [comment]
Is (1) a small change? I'm not sure I can fully visualize the impact but if it's simple to show in a PR then let's just take a look

**Comment by Gabriel-Ladzaretti** [comment]
> Is (1) a small change? I'm not sure I can fully visualize the impact but if it's simple to show in a PR then let's just take a look

This is all there is to it. 
 - https://github.com/Gabriel-Ladzaretti/renovate/pull/7

now we cache the branch and can utilize the repository cache within the migration flow.

**Comment by rarkins** [comment]
That looks like an elegant solution to me. Do you have any concerns or doubts about it?

**Comment by Gabriel-Ladzaretti** [comment]
> That looks like an elegant solution to me. Do you have any concerns or doubts about it?

Currently the only thing i can think of is that this flow might be run more than once in a current run due to the recursive call to `renovateRepository`.

that said, thats not much different than running self hosted with cache disable, what i mean is that we detect that the branch is not modified and we basically skip it. same goes for cache, if we utilize it correctly this shouldnt be a problem.

other than that the logic stays the same, we just change execution point

**Comment by rarkins** [comment]
Recursive call within Renovate OSS or within the app's wrapper?

**Comment by Gabriel-Ladzaretti** [comment]
OSS -

declaration -
https://github.com/renovatebot/renovate/blob/4f82df4c3d501641ae05ce713bb1328d339ec5d9/lib/workers/repository/index.ts#L21-L25

recursive call - 
https://github.com/renovatebot/renovate/blob/4f82df4c3d501641ae05ce713bb1328d339ec5d9/lib/workers/repository/index.ts#L54-L60

rubber ducking, seems that the recursive depth is 2 at most due to the `canRetry` arg

**Comment by viceice** [comment]
and it's only done on automerge

**Comment by Gabriel-Ladzaretti** [comment]
The base-branch SHA comparison wont help us much here. I've tested this approach and I think it won't work, as the `parentSha` saved for a given branch is the SHA at the moment of commit to the new branch. 
Therefore, all it takes to be out of sync in this approach is a single commit to the base branch.
This obviously happens quite a bit, while the config is unchanged most of the time, but in this scenario we will try to migrate on every run which will lead for some extra api/fs work.
  
A possible solution might be in the form of 

 - https://github.com/renovatebot/renovate/pull/17021

as the migration flow is already dependent on `detectRepoFileConfig` which itself fetches the config, therefore caching and returning an additional raw data from it saves us doing so again and we are not reliant on the repo cache at all. 

This is done regardless of repo cache.

**Comment by Gabriel-Ladzaretti** [comment]
superseded by
 - https://github.com/renovatebot/renovate/pull/17311
